To Do: Implementation – AdaBoost

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score

data = pd.read_csv("income.csv")

data = data.dropna()

le = LabelEncoder()
for col in data.columns:
    if data[col].dtype == 'object':
        data[col] = le.fit_transform(data[col])

X = data.iloc[:, :-1]   
y = data.iloc[:, -1]    

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

model_default = AdaBoostClassifier(n_estimators=10, random_state=42)
model_default.fit(X_train, y_train)

y_pred_default = model_default.predict(X_test)
accuracy_default = accuracy_score(y_test, y_pred_default)

print("Accuracy with default n_estimators (10):", accuracy_default)

estimators_list = [5, 10, 20, 50, 100]

best_score = 0
best_n = 0

print("\nTuning Results:")

for n in estimators_list:
    model = AdaBoostClassifier(n_estimators=n, random_state=42)
    model.fit(X_train, y_train)
    
    y_pred = model.predict(X_test)
    acc = accuracy_score(y_test, y_pred)
    
    print(f"Trees: {n}  Accuracy: {acc}")
    
    if acc > best_score:
        best_score = acc
        best_n = n

print("\nBest Number of Trees:", best_n)
print("Best Accuracy:", best_score)

Accuracy with default n_estimators (10): 0.8182004299314157

Tuning Results:
Trees: 5  Accuracy: 0.8100112601085065
Trees: 10  Accuracy: 0.8182004299314157
Trees: 20  Accuracy: 0.8244446719213839
Trees: 50  Accuracy: 0.8327362063670796
Trees: 100  Accuracy: 0.8328385709898659

Best Number of Trees: 100
Best Accuracy: 0.8328385709898659


In [2]:
from sklearn.metrics import confusion_matrix

# After best model prediction
best_model = AdaBoostClassifier(n_estimators=best_n, random_state=42)
best_model.fit(X_train, y_train)

y_pred_best = best_model.predict(X_test)

cm = confusion_matrix(y_test, y_pred_best)
print("\nConfusion Matrix:\n", cm)


Confusion Matrix:
 [[7117  297]
 [1336 1019]]


After building the AdaBoost model, write the answer for the following question in your observation book
1. For “income.csv” dataset,
What is the best accuracy score and confusion matrix of the classifier you observed and using how many trees ?

Best Accuracy Score: 0.83283

Best Number of Trees: 100

Confusion Matrix:

[[7117     297]

 [1336     1019]]